solutions/gold/baltic-09-candy.mdx

```diff
@@ -0,0 +1,150 @@
+---
+id: baltic-09-candy
+source: Baltic OI 2009
+title: Candy Machine
+author: Panos Paskalis
+---
+
+[Official Analysis](http://www.csc.kth.se/contest/boi/candy-spoiler.pdf)
+
+## Explanation
+
+### Observation
+
+If there is a candy $i$ $(s_i, t_i)$ and a candy $j$ $(s_j, t_j)$, with
+$t_i < t_j$, then $j$ can be caught by the same wagon as $i$, if the distance
+between them can be covered by the wagon in $t_j - t_i$ time. That is:
+
+$$
+|s_j - s_i| \leq t_j - t_i \iff
+t_i - t_j \leq s_j - s_i \leq t_j - t_i \iff
+$$
+$$
+\begin{cases}
+    s_i + t_i \leq s_j + t_j \text{ and } \\
+    s_j - t_j \leq s_i - t_i \\
+\end{cases}
+$$
+Since $t_i \geq 0$,
+$$
+s_i - t_i \leq s_i + t_i
+$$
+In total:
+$$
+s_j - t_j \leq s_i - t_i \leq s_i + t_i \leq s_j + t_j
+$$
+
+### Finding the minimum number of wagons
+
+If we think of each candy as a segment $[s - t, s + t]$
+(note that each candy corresponds to a unique segment and vice versa), we can
+conclude that then candy $j$ can be caught by the same wagon as candy $i$ if
+the segment of $i$ is nested inside the segment of $j$.
+Therefore a wagon can catch a sequence of candies $c_1, c_2, ..., c_k$
+if they form a sequence of nested segments. It follows that the minimum number
+of wagons needed is the longest sequence of non-nested segments.
+
+Two segments $[l_1, r_1]$, $[l_2, r_2]$ that are not nested adhere to:
+$$l_1 < l_2 \implies r_1 < r_2$$.
+
+If we sort the candies by the left endpoint of their segments such that
+$$
+l_1 \leq l_2 \leq ... \leq l_n
+$$
+then a sequence of non-nested segments would correspond to an increasing
+sequence of right endpoints. Therefore, the minimum number of required wagons
+would correspond to the length of the longest increasing sequence of right
+endpoints.
+
+### Assigning the candies to wagons
+
+By following the binary search algorithm described in
+[the LIS module](/gold/lis), we can assign the candies with the same LIS
+length in the same wagon. The segments that have the same LIS length cannot be
+nested. If they were, we could use the inside segment to increase the LIS of
+the outside one by 1.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+<Warning>
+Watch out for the case $l_i = l_j$, in which case we must sort by decreasing
+right endpoint to ensure they would not be counted as non-nested segments.
+</Warning>
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+typedef pair<int, int> pi;
+
+const int MAXN = 100000;
+
+int n;
+pi candies[MAXN];
+int lis[MAXN + 1];
+vector<pi> candies_caught[MAXN];
```

Comment:
these have no need to be global
also please just use vectors & size them based on `n`

Reply:
Done.